New rOCCI client API and DSL

arax opened the pull request:
## Changes:
- new rOCCI client API, added helpers, rewrote parts of the code, made it more user-friendly in general
- new rOCCI client DSL, very simple, basically just a wrapper for the API
- updated rOCCI client examples (x509auth + DSL)
## TODO:
- rOCCI API/DSL documentation
- rOCCI client executables
- rOCCI client tests

arax committed:
First draft of the new client API

Only list and describe are working at the moment.

arax committed:
Added mixin helpers, basic create functionality

arax committed:
Added basic delete functionality

rOCCI client should now be able to list, describe, create
and delete resources. This has been tested against the current
version of the rOCCI-server (i.e., v0.5). Support for
VM instantiation from native ON templates should also work.

arax committed:
Fixed code indentation

arax committed:
Simplified client API, minor fixes and comments

arax committed:
Added first draft of the client DSL

arax committed:
Added DSL example, still WIP

arax committed:
Fixed a bug and a typo in the DSL

DSL method 'mixins' has been renamed to 'mixinz' to avoid
a strange bug causing mixups in lib/occi/collection.rb
(on line 23).

arax committed:
Fixed code formatting in dsl.rb

arax committed:
Added a skeleton for the occi executable

This commit includes minor changes in the directory
structure. It should help us separate API files from
the actual client (bin/occi) lib files in the future.

ffeldhaus commented:
You did a great job with creating the DSL and improving the client.

Some comments:
- why did you name mixins mixinz in the DSL?
- I like the way you are handling human readable resource types. I would suggest though, that it should be more dynamic based on the model provided by the service provider. We could use the term of a kind as human readable name to map it to the type identifier. Terms of kinds should be unique inside a service provider.
- Why did you use get_instance instead of get_resource? Which would be more intuitive?
- it might be a good idea to add get_mixin_type_identifiers
- if we would use the term as human readable name of kinds, then we would not need get_resource_type or at least could replace it with a simple lookup of the kinds term

arax commented:
- why did you name mixins mixinz in the DSL?

It was causing this very strange error https://gist.github.com/3613751 so I had to rename it, at least for now
- I like the way you are handling human readable resource types. I would suggest though, that it should be more dynamic based on the model provided by the service provider. We could use the term of a kind as human readable name to map it to the type identifier. Terms of kinds should be unique inside a service provider.

That's a good idea, if their behavior is the same, i.e. I can list/describe/create/delete them using the same approach (and the same method). I wasn't sure what can and cannot be a kind. For instance, is infrastructure#storagelink or infrastructure#networkinterface kind too? What about core#resource? They are all listed as kinds when I open /-/. 
- Why did you use get_instance instead of get_resource? Which would be more intuitive?

get_resource and resource in DSL would be more intuitive.
- it might be a good idea to add get_mixin_type_identifiers

It just slipped my mind.
- if we would use the term as human readable name of kinds, then we would not need get_resource_type or at least could replace it with a simple lookup of the kinds term

Ok.

arax committed:
Improved client API and DSL

*Renamed mixinz back to mixins (using extend instead of include)
*Renamed instance to resource (more intuitive)
*Added get_mixin_type_identifiers to API and mixin_type_identifiers to
DSL
